### ISSUE TITLE
Embedding Dimension: Up to 1024, supports user-defined output dimensions ranging from 32 to 1024

### ISSUE THREAD

**Issue by datalee** [QwenLM/Qwen3-Embedding#2]
[32, 1024]的任何整数都行吗？
我看别人都是32的倍数

![Image](https://github.com/user-attachments/assets/1bb0224d-3570-4ea4-bf46-05cb1b7a4a86)

**Comment by datalee** [comment]
> 是的，支持任意维度

有调用代码示例吗？怎么设置，我看sentence transformer里有个参数：truncate_dim
`from sentence_transformers import SentenceTransformer

model = SentenceTransformer("tomaarsen/static-similarity-mrl-multilingual-v1", truncate_dim=256)
embeddings = model.encode([
    "I used to hate him.",
    "Раньше я ненавидел его."
])
print(embeddings.shape)
# => (2, 256)`

**Comment by datalee** [comment]
懂了，加个示例：
`
import torch
import torch.nn.functional as F

from torch import Tensor
from transformers import AutoTokenizer, AutoModel


def last_token_pool(last_hidden_states: Tensor,
                 attention_mask: Tensor) -> Tensor:
    left_padding = (attention_mask[:, -1].sum() == attention_mask.shape[0])
    if left_padding:
        return last_hidden_states[:, -1]
    else:
        sequence_lengths = attention_mask.sum(dim=1) - 1
        batch_size = last_hidden_states.shape[0]
        return last_hidden_states[torch.arange(batch_size, device=last_hidden_states.device), sequence_lengths]


def get_detailed_instruct(task_description: str, query: str) -> str:
    return f'Instruct: {task_description}\nQuery:{query}'

def tokenize(tokenizer, input_texts, eod_id, max_length):
    batch_dict = tokenizer(input_texts, padding=False, truncation=True, max_length=max_length-2)
    for seq, att in zip(batch_dict["input_ids"], batch_dict["attention_mask"]):
        seq.append(eod_id)
        att.append(1)
    batch_dict = tokenizer.pad(batch_dict, padding=True, return_tensors="pt")
    return batch_dict

# Each query must come with a one-sentence instruction that describes the task
task = 'Given a web search query, retrieve relevant passages that answer the query'

queries = [
    get_detailed_instruct(task, 'What is the capital of China?'),
    get_detailed_instruct(task, 'Explain gravity')
]
# No need to add instruction for retrieval documents
documents = [
    "The capital of China is Beijing.",
    "Gravity is a force that attracts two bodies towards each other. It gives weight to physical objects and is responsible for the movement of planets around the sun."
]
input_texts = queries + documents

tokenizer = AutoTokenizer.from_pretrained(r'D:\backup\Qwen3-Embedding-0.6B', padding_side='left')
model = AutoModel.from_pretrained(r'D:\backup\Qwen3-Embedding-0.6B')

# We recommend enabling flash_attention_2 for better acceleration and memory saving.
# model = AutoModel.from_pretrained('Qwen/Qwen3-Embedding-0.6B', attn_implementation="flash_attention_2", torch_dtype=torch.float16).cuda()

eod_id = tokenizer.convert_tokens_to_ids("<|endoftext|>")
max_length = 8192

# Tokenize the input texts
batch_dict = tokenize(tokenizer, input_texts, eod_id, max_length)
batch_dict.to(model.device)
outputs = model(**batch_dict)
embeddings = last_token_pool(outputs.last_hidden_state, batch_dict['attention_mask'])

# MRL
#  You can extract the unnormalized embedding from output and 
#  manually adjust (truncate or project) the embedding dimensions as needed.
embeddings = embeddings[:, :128]

# normalize embeddings
embeddings = F.normalize(embeddings, p=2, dim=1)
scores = (embeddings[:2] @ embeddings[2:].T)
print(scores.tolist())
`

**Comment by yanfan0531** [comment]
How do I use a truncated embedding via SentenceTransformer?

**Comment by datalee** [comment]
> How do I use a truncated embedding via SentenceTransformer? Does the param truncate_dim works before normalization?


it's ok。
```python
from sentence_transformers import SentenceTransformer
model = SentenceTransformer(r'D:\backup\Qwen3-Embedding-0.6B', truncate_dim=128)

# We recommend enabling flash_attention_2 for better acceleration and memory saving,
# together with setting `padding_side` to "left":
# model = SentenceTransformer(
#     "Qwen/Qwen3-Embedding-0.6B",
#     revision="refs/pr/2",
#     model_kwargs={"attn_implementation": "flash_attention_2", "device_map": "auto"},
#     tokenizer_kwargs={"padding_side": "left"},
# )

# The queries and documents to embed
queries = [
    "What is the capital of China?",
    "Explain gravity",
]
documents = [
    "The capital of China is Beijing.",
    "Gravity is a force that attracts two bodies towards each other. It gives weight to physical objects and is responsible for the movement of planets around the sun.",
]

# Encode the queries and documents. Note that queries benefit from using a prompt
# Here we use the prompt called "query" stored under `model.prompts`, but you can
# also pass your own prompt via the `prompt` argument
query_embeddings = model.encode(queries, prompt_name="query")
document_embeddings = model.encode(documents)

# Compute the (cosine) similarity between the query and document embeddings
similarity = model.similarity(query_embeddings, document_embeddings)
print(similarity)
```

**Comment by tianchengcheng-cn** [comment]
> > > 懂了，加个示例：
> > > import torch
> > > import torch.nn.functional as F
> > > from torch import Tensor
> > > from transformers import AutoTokenizer, AutoModel
> > > def last_token_pool(last_hidden_states: Tensor,
> > > attention_mask: Tensor) -> Tensor:
> > > left_padding = (attention_mask[:, -1].sum() == attention_mask.shape[0])
> > > if left_padding:
> > > return last_hidden_states[:, -1]
> > > else:
> > > sequence_lengths = attention_mask.sum(dim=1) - 1
> > > batch_size = last_hidden_states.shape[0]
> > > return last_hidden_states[torch.arange(batch_size, device=last_hidden_states.device), sequence_lengths]
> > > def get_detailed_instruct(task_description: str, query: str) -> str:
> > > return f'Instruct: {task_description}\nQuery:{query}'
> > > def tokenize(tokenizer, input_texts, eod_id, max_length):
> > > batch_dict = tokenizer(input_texts, padding=False, truncation=True, max_length=max_length-2)
> > > for seq, att in zip(batch_dict["input_ids"], batch_dict["attention_mask"]):
> > > seq.append(eod_id)
> > > att.append(1)
> > > batch_dict = tokenizer.pad(batch_dict, padding=True, return_tensors="pt")
> > > return batch_dict
> > > # Each query must come with a one-sentence instruction that describes the task
> > > task = 'Given a web search query, retrieve relevant passages that answer the query'
> > > queries = [
> > > get_detailed_instruct(task, 'What is the capital of China?'),
> > > get_detailed_instruct(task, 'Explain gravity')
> > > ]
> > > # No need to add instruction for retrieval documents
> > > documents = [
> > > "The capital of China is Beijing.",
> > > "Gravity is a force that attracts two bodies towards each other. It gives weight to physical objects and is responsible for the movement of planets around the sun."
> > > ]
> > > input_texts = queries + documents
> > > tokenizer = AutoTokenizer.from_pretrained(r'D:\backup\Qwen3-Embedding-0.6B', padding_side='left')
> > > model = AutoModel.from_pretrained(r'D:\backup\Qwen3-Embedding-0.6B')
> > > # We recommend enabling flash_attention_2 for better acceleration and memory saving.
> > > # model = AutoModel.from_pretrained('Qwen/Qwen3-Embedding-0.6B', attn_implementation="flash_attention_2", torch_dtype=torch.float16).cuda()
> > > eod_id = tokenizer.convert_tokens_to_ids("<|endoftext|>")
> > > max_length = 8192
> > > # Tokenize the input texts
> > > batch_dict = tokenize(tokenizer, input_texts, eod_id, max_length)
> > > batch_dict.to(model.device)
> > > outputs = model(**batch_dict)
> > > embeddings = last_token_pool(outputs.last_hidden_state, batch_dict['attention_mask'])
> > > # MRL
> > > # You can extract the unnormalized embedding from output and
> > > # manually adjust (truncate or project) the embedding dimensions as needed.
> > > embeddings = embeddings[:, :128]
> > > # normalize embeddings
> > > embeddings = F.normalize(embeddings, p=2, dim=1)
> > > scores = (embeddings[:2] @ embeddings[2:].T)
> > > print(scores.tolist())
> > 
> > 
> > 你这样直接截断太残暴了，会损失很多信息吧。感觉不是这么暴力的
> 
> 事实就是如此，可以去看看论文（Matryoshka Representation Learning：[https://arxiv.org/abs/2205.13147）](https://arxiv.org/abs/2205.13147%EF%BC%89)
> 
> torch截断和SentenceTransformer的truncate_dim做了对比，结果是一致的。 我上面代码的结果（用的demo示例）
> 
> ```
> # 128 dim
> # [[0.7877246737480164, 0.21917583048343658], [0.18308258056640625, 0.6724958419799805]]
> ```
> 
> 用SentenceTransformer参数的结果：
> 
> ```
> # 128 dim
> tensor([[0.7877, 0.2192],
>         [0.1831, 0.6725]])
> ```

但MRL人家是要训练的，然后经过调整损失，保证前缀截取的结果保留更多的精度信息。但原版的qwen-embedding没有这一部分吧

**Comment by datalee** [comment]
> > > > 懂了，加个示例：
> > > > import torch
> > > > import torch.nn.functional as F
> > > > from torch import Tensor
> > > > from transformers import AutoTokenizer, AutoModel
> > > > def last_token_pool(last_hidden_states: Tensor,
> > > > attention_mask: Tensor) -> Tensor:
> > > > left_padding = (attention_mask[:, -1].sum() == attention_mask.shape[0])
> > > > if left_padding:
> > > > return last_hidden_states[:, -1]
> > > > else:
> > > > sequence_lengths = attention_mask.sum(dim=1) - 1
> > > > batch_size = last_hidden_states.shape[0]
> > > > return last_hidden_states[torch.arange(batch_size, device=last_hidden_states.device), sequence_lengths]
> > > > def get_detailed_instruct(task_description: str, query: str) -> str:
> > > > return f'Instruct: {task_description}\nQuery:{query}'
> > > > def tokenize(tokenizer, input_texts, eod_id, max_length):
> > > > batch_dict = tokenizer(input_texts, padding=False, truncation=True, max_length=max_length-2)
> > > > for seq, att in zip(batch_dict["input_ids"], batch_dict["attention_mask"]):
> > > > seq.append(eod_id)
> > > > att.append(1)
> > > > batch_dict = tokenizer.pad(batch_dict, padding=True, return_tensors="pt")
> > > > return batch_dict
> > > > # Each query must come with a one-sentence instruction that describes the task
> > > > task = 'Given a web search query, retrieve relevant passages that answer the query'
> > > > queries = [
> > > > get_detailed_instruct(task, 'What is the capital of China?'),
> > > > get_detailed_instruct(task, 'Explain gravity')
> > > > ]
> > > > # No need to add instruction for retrieval documents
> > > > documents = [
> > > > "The capital of China is Beijing.",
> > > > "Gravity is a force that attracts two bodies towards each other. It gives weight to physical objects and is responsible for the movement of planets around the sun."
> > > > ]
> > > > input_texts = queries + documents
> > > > tokenizer = AutoTokenizer.from_pretrained(r'D:\backup\Qwen3-Embedding-0.6B', padding_side='left')
> > > > model = AutoModel.from_pretrained(r'D:\backup\Qwen3-Embedding-0.6B')
> > > > # We recommend enabling flash_attention_2 for better acceleration and memory saving.
> > > > # model = AutoModel.from_pretrained('Qwen/Qwen3-Embedding-0.6B', attn_implementation="flash_attention_2", torch_dtype=torch.float16).cuda()
> > > > eod_id = tokenizer.convert_tokens_to_ids("<|endoftext|>")
> > > > max_length = 8192
> > > > # Tokenize the input texts
> > > > batch_dict = tokenize(tokenizer, input_texts, eod_id, max_length)
> > > > batch_dict.to(model.device)
> > > > outputs = model(**batch_dict)
> > > > embeddings = last_token_pool(outputs.last_hidden_state, batch_dict['attention_mask'])
> > > > # MRL
> > > > # You can extract the unnormalized embedding from output and
> > > > # manually adjust (truncate or project) the embedding dimensions as needed.
> > > > embeddings = embeddings[:, :128]
> > > > # normalize embeddings
> > > > embeddings = F.normalize(embeddings, p=2, dim=1)
> > > > scores = (embeddings[:2] @ embeddings[2:].T)
> > > > print(scores.tolist())
> > > 
> > > 
> > > 你这样直接截断太残暴了，会损失很多信息吧。感觉不是这么暴力的
> > 
> > 
> > 事实就是如此，可以去看看论文（Matryoshka Representation Learning：[https://arxiv.org/abs/2205.13147）](https://arxiv.org/abs/2205.13147%EF%BC%89)
> > torch截断和SentenceTransformer的truncate_dim做了对比，结果是一致的。 我上面代码的结果（用的demo示例）
> > ```
> > # 128 dim
> > # [[0.7877246737480164, 0.21917583048343658], [0.18308258056640625, 0.6724958419799805]]
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 用SentenceTransformer参数的结果：
> > ```
> > # 128 dim
> > tensor([[0.7877, 0.2192],
> >         [0.1831, 0.6725]])
> > ```
> 
> 但MRL人家是要训练的，然后经过调整损失，保证前缀截取的结果保留更多的精度信息。但原版的qwen-embedding没有这一部分吧

有，他们工程师在一次直播分享里提到了，但是没在paper里写，这就很炒蛋

**Comment by Cgetier520990** [comment]
看到您说懂了，是采用这种截断的方式吗？
这种截断的向量与原来的准确性相差多少？比如说，原来 4B 的是 2560，现在截断到 1024，准确性会不会降低
# MRL
#  You can extract the unnormalized embedding from output and 
#  manually adjust (truncate or project) the embedding dimensions as needed.
embeddings = embeddings[:, :128]
